crates/sargon/src/gateway_api/models/types/response/state/entity/details/details/item.rs

```diff
@@ -0,0 +1,251 @@
+use crate::prelude::*;
+
+#[derive(Deserialize, Serialize, Clone, PartialEq, Eq, Debug, EnumAsInner)]
+#[serde(tag = "type")]
+pub enum StateEntityDetailsResponseItemDetails {
+    FungibleResource(StateEntityDetailsResponseFungibleResourceDetails),
+    NonFungibleResource(StateEntityDetailsResponseNonFungibleResourceDetails),
+    FungibleVault,
+    NonFungibleVault,
+    Package(StateEntityDetailsResponsePackageDetails),
+    Component(StateEntityDetailsResponseComponentDetails),
+}
+
+impl StateEntityDetailsResponseItemDetails {
+    /// Returns true if the entity can be transferred.
+    ///
+    /// To determine whether an entity can be transferred, we need to check its assignments
+    /// for the ("Main", "depositor") & ("Main", "withdrawer") roles.
+    /// Once we have them, we first check if the assignments are `Explicit` or `Owner`.
+    /// - If they are `Explicit`, the entity will be transferable if both the depositor & withdrawer
+    /// have `AllowAll` rules.
+    /// - If they are `Owner`, the entity will be transferable if the owner has `AllowAll` rule.
+    pub fn can_be_transferred(&self) -> bool {
+        let Some(role_assignments) = self.role_assignments() else {
+            return false;
+        };
+        let Some(depositor) = role_assignments
+            .entries
+            .clone()
+            .into_iter()
+            .find(|entry| entry.role_key == RoleKey::main_depositor())
+        else {
+            return false;
+        };
+
+        if depositor.assignment.resolution == RoleAssignmentResolution::Owner {
+            // No need to check for withdrawer, as its `resolution` will be the same.
+            return role_assignments.owner.rule == ExplicitRule::AllowAll;
+        }
+
+        let Some(withdrawer) = role_assignments
+            .entries
+            .into_iter()
+            .find(|entry| entry.role_key == RoleKey::main_withdrawer())
+        else {
+            return false;
+        };
+
+        depositor.assignment.explicit_rule == Some(ExplicitRule::AllowAll)
```

Comment:
I prefer naming logical condition and build up logical expression from two named conditions. Much easier to read / maintain / doc.

```rust
let allows_all = |role_assignment: &ComponentEntityRoleAssignments | {
    role_assignment .assignment.explicit_rule == Some(ExplicitRule::AllowAll)
};
let depositor_allows_all = allows_all(&depositor);
let withdrawer_allows_all = allows_all(&withdrawer);
// Both depositor and withdrawer must allow
let can_be_transferred = depositor_allows_all && withdrawer_allows_all;
#[allow(clippy::something_something_needless_assign]
can_be_transferred
```